### DISCUSSION TITLE
Get crop phenology stage dates in report

### DISCUSSION THREAD

**Discussion by LenLon** [APSIMInitiative/ApsimX#9192] [Q&A] [answered]
Hello,

I want to report out the dates when a crop is entering a new phenological stage (emergence, flowering, grain fill, harvest) using only a single report event, so something akin to the tracker function in Apsim Classic. This would also enable tracking maxT during flowering and other agroclimatic indices very useful for building predictors for ML-hybrid models. 

I was able to do this in Classic using statements like

```
value of maize.DaysAfterSowing on maize.end_of_juvenile as EJ_Days
value of maize.DaysAfterSowing on maize.flowering as FL_Days
```

but there seems to be no `value of  x on y` aggregation function anymore in ApsimX. Is this still possible using the report node or would this require a manager script? 

I can get agroclimatic indices to some degree in ApsimX already using statements like 
`max of [Weather].MaxT from [Maize].Flowering to [Maize].Harvesting as maxT_flower`
but apart from sowing, flowering and harvesting there seem to be no other event options for (maize) phenological phases, so maybe this is also a report variable feature request.

**Comment by hol353** [comment]
We should probably add 'value of ... on ...'  functionality.

As a workaround, you might be able to use a sum function like this:

```sum of [Maize].DaysAfterSowing from [Maize].Flowering to [Maize].Flowering as FL_Days```

The next problem is that not all phenology events can be trapped by report so you probably can't do end juvenile. To solve this problem, #8095 needs to be fixed.

You best bet might be to write a manager script, which isn't straight forward.

**Reply by LenLon** [reply]
Okay, I hope that the event.publish capability is implemented soon as I'm no good with writing these manager scripts without help from a dev. :'-( 
Is there a flag for when a phenology phase starts which I could parse to Clock.Today or DateUtilities? I wouldn't even need the name of the phenological phase, just a date where phenology enters a new one. This could be really useful if the script manages to be somewhat crop-agnostic, so it'd work for any or most crops in ApsimX.

**Reply by hol353** [reply]
We won't get to event publish anytime soon unfortunately.

**Comment by hut104** [comment]
I often just report all phenology days after sowing values at harvest.


Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by LenLon** [reply]
Interesting - do you mean the days after sowing values for each phenological phase? How do you go about that?

**Comment by hol353** [comment]
Another option is to simply output daily and then process the output file in EXCEL or R or something else.

**Reply by LenLon** [reply]
True, but this balloons output data 365-fold, which makes handling a lot slower - not ideal if you are short on computing time and space!

**Comment by LenLon** [comment]
I have tinkered a lot with the manager in the last days, and have some sort of prototype manager script for this task set up:

```
using Models.PMF;
using Models.Core;
using System;


namespace Models
{
    [Serializable] 
    public class Script : Model
    {
    	[Link] private Clock Clock;
    	
        [Description("Crop")]
        public Plant Crop { get; set; }
        
        public bool stage_sow { get; set; } = false;
        
        public string sow_date { get; set; }
        
        public bool stage_florI { get; set; } = false;
        
        public bool stage_flower { get; set; } = false;
        
        public bool stage_startgr { get; set; } = false;
        
        [EventSubscribe("DoManagement")]
        private void OnDoManagement(object sender, EventArgs e)
        {
        	if (Crop.Phenology.CurrentStageName == "FloralInitiation")
        	{
        		stage_florI = true;
        	}
        	else if (Crop.Phenology.CurrentStageName == "Flowering")
        	{
        		stage_flower = true;
        	}
        	else if (Crop.Phenology.OnStartDayOf("StartGrainFill"))
        	{
        		stage_startgr = true;
        	}
        	else if (Crop.Phenology.OnStartDayOf("Sowing"))
        	{
        		stage_sow = true;
        		sow_date = Clock.Today.ToString();
        	}
        	else
        	{
        		stage_sow = false;
        		stage_florI = false;
        		stage_flower = false;
        		stage_startgr = false;
        	}
        }
    }
}
```

This compiles, but only works for the "Sowing" phenology stage. If I add the variables to my report using `[CropStatus1].ScriptSimulationsSimulationFieldCropStatus1.stage_sow as sow_check` with a daily report frequency, I can see this:

![image](https://github.com/user-attachments/assets/0c1d86dd-ac1b-44b3-88ae-35c3a8035c1e)

So far so good. However, when report out the other stages (floral initiation, flowering etc.) they never evaluate to true, as evidenced here:

![image](https://github.com/user-attachments/assets/cbbc163d-3c45-4d3b-8ccd-2c54d7835cca)

I am quite proud I even made it this far, but now I am stumped. The same problem exists if I use `Crop.Phenology.CurrentPhaseName` instead, it works for "Germinating" but not for any other phase. I thought I found a clever way using `Crop.Phenology.OnStartDayOf`, but that seems to make no difference. The end goal is to trigger reporting on changes in phenology stages, using the nice report events of syntax `(bool)CropStatus1.GetProperty("stage_startgr") == true` :-)

**Comment by hol353** [comment, marked as answer]
Well done on getting this far! Many users don't.

Turns out it is a within day ordering issue in your manager script. Your manager code will be called close to the start of the day. At that time the phenology model hasn't calculated it's variables yet for the day. You need to have your manager code called after the other models have all calculated their variables. Change the EventSubscribe line to this:

```c#
[EventSubscribe("DoManagementCalculations")]
```

**Reply by LenLon** [reply]
This worked, thanks so much! 

It would be handy to have a list of events the manager can subscribe to in order of when they are called (and especially what happens when they are called), so far I've got

```
StartOfSimulation 
DoManagement 
DoManagementCalculations 
```

Are these all of them or are there others?
